### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by SanjoDeundiak**

### CODE REVIEW

**Comment by SanjoDeundiak** [comment]
Work to be done:
FIXME&TODOS
Add hashes other than SHA256
Review input/output formats (little endian/big endian, etc.)
Sanity checks (especially on data provided from user)
Multithreading?
Optimization (prevent obsolete copying, etc.)
Refactoring
Switch dependencies to stable branches (franklin crypto, etc.)
Compare my implementation with latest zksync master
More tests
Docs

**Comment by dvush** [comment]
Hi, please, can you run `cargo fmt` so code will be formatted for ease of reading.
Also, you can lint it using `cargo clippy`.

**Comment by SanjoDeundiak** [comment]
@dvush done

File: .gitignore
Comment (popzxc) [inline .gitignore]: Nit: Please add the missing endlines at the end of files here and in other places.

File: Cargo.toml
Comment (popzxc) [inline Cargo.toml]: Please split this crate into two different crates:

1. With the multisig implementation, designed to be used as a dependency in rust projects.
2. With the WASM bindings (which uses 1st crate as a dependency).

File: src/musig.rs
Comment (popzxc) [inline src/musig.rs]: Please add a doc-comment to this structure (and in other places where it makes sense) which explains the purpose and basic usage.

File: src/musig.rs
Comment (popzxc) [inline src/musig.rs]: I'd suggest to use template parameters here to avoid dynamic dispatch.

File: src/musig.rs
Comment (popzxc) [inline src/musig.rs]: Nit: It is preferable to refer to the type in the implementation as `Self` rather than the direct type name.

File: README.md
Comment (popzxc) [inline README.md]: Nit: Correct shortcut for Rust syntax highlight in markdown is `rust`, not `rs`.

File: src/musig.rs
Comment (popzxc) [inline src/musig.rs]: Please either elaborate what kind of optimization should be done or resolve this TODO.

File: src/musig.rs
Comment (popzxc) [inline src/musig.rs]: `&dyn` is not really required here (`impl AggregateHash<E>` should be sufficient), but I think using the template parameters will resolve this problem.

File: src/musig.rs
Comment (popzxc) [inline src/musig.rs]: Do I understand it correctly that `r` here is not used as a `PrivateKey`, but rather as a workaround for a multiplication?
If so, it worsens the readability of the code, as it conflicts with the implied type semantics.

File: src/musig.rs
Comment (popzxc) [inline src/musig.rs]: Is there a reason to return a `&Vec<u8>` rather than `&[u8]`?

File: src/musig.rs
Comment (popzxc) [inline src/musig.rs]: Data related to `self.self_index` is accessed multiple times and it is considered an invariant that this value is never `None`. Let's enforce it on the language level and store the self-related data as the actual values rather than `Option`.

File: src/musig_hash.rs
Comment (popzxc) [inline src/musig_hash.rs]: Are all the trait types required? Is it possible that one will want to use `AggregateHash` but totally not interested in `CommitmentHash` in the context of this library?

File: src/musig_hash.rs
Comment (popzxc) [inline src/musig_hash.rs]: Nit: Missing `Debug` implementation. It's a good practice to implement `Debug` (either by deriving or manually) to any type, so it'll be easier to debug the code later. Actually there even is a [lint](https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#missing-debug-implementations) for this.

File: src/musig_hash.rs
Comment (popzxc) [inline src/musig_hash.rs]: `32` is a magic number, it's better to provide a constant for it.

File: src/musig_hash.rs
Comment (popzxc) [inline src/musig_hash.rs]: Do I understand it correctly that `pubs` list in this method is expected to be the same, and only `last` part differs from call to call? If so, wouldn't it make more sense to cache the part of calculation to avoid doing the same thing?

File: src/musig_hash.rs
Comment (popzxc) [inline src/musig_hash.rs]: Is it possible to avoid dynamic allocation by introducing a hash result type, which will hold its contents on the stack?

File: src/musig_hash.rs
Comment (popzxc) [inline src/musig_hash.rs]: Can we simply use `Sha256::digest(m).to_vec()` here?

File: src/musig_wasm.rs
Comment (popzxc) [inline src/musig_wasm.rs]: Currently this file is pretty big and contains a lot of different types/functions. Wouldn't it be better to split it into several logically-separated files?

File: src/musig_wasm.rs
Comment (popzxc) [inline src/musig_wasm.rs]: Nit: I believe you can elide the explicit type here by invoking the method on the type itself rather than on trait.

File: src/musig_wasm.rs
Comment (popzxc) [inline src/musig_wasm.rs]: It is a good practice to use the `Vec::with_capacity` method in the situations when the expected output size is known.

File: tests/musig_tests.rs
Comment (popzxc) [inline tests/musig_tests.rs]: You don't have to specify `#[cfg(test)]` here, since it's a `tests` directory which is only compiled if this attribute is provided.

File: tests/musig_tests.rs
Comment (popzxc) [inline tests/musig_tests.rs]: You can simply use `while let Some(session) = iterator.next()` or even `for session in iterator`

File: tests/musig_tests.rs
Comment (popzxc) [inline tests/musig_tests.rs]: I'd prefer to get rid of the unsafe code, as it doesn't introduce a significant improvements in either performance or readability. As far as I can see, this method is used to iterate through the list skipping exactly one element in it, and it can easily (and as efficient) be done with the safe code.

File: tests/musig_tests.rs
Comment (popzxc) [inline tests/musig_tests.rs]: Personally I'd love to see more tests for this crate, both unit and integration ones. As a simple example, it will be nice to have a test that tries to verify an incorrect signature.

File: tests/musig_wasm_tests.rs
Comment (popzxc) [inline tests/musig_wasm_tests.rs]: I'm not sure if these tests are actually useful: as the WASM part simply delegates the input to the Rust implementation of the algorithm, it is a bit overkill to test the whole logic again. It should be sufficient to check that the `MusigSession` and `MusigWasm` operate the same way given the same input. 

However, I think that it will be useful to check that wrapper methods do behave correctly with an unexpected input.

File: src/lib.rs
Comment (popzxc) [inline src/lib.rs]: `extern crate` statements aren't required in the 2018 edition.

File: tests/musig_tests.rs
Comment (popzxc) [inline tests/musig_tests.rs]: You can assign a result of `if` statements in Rust:

```rust
let value = if true {
    42
} else {
    24
};
```

**Review by popzxc — CHANGES_REQUESTED** [review]
First of all, thanks for your work! In general, it looks good.

All the comments in this review are subjects for discussion, if you don't agree with some of them, do not hesitate to defend your position :)

Comments marked as "Nit" are minor/subjective and not mandatory to implement.

File: src/musig_hash.rs
Comment (shamatar) [inline src/musig_hash.rs]: This function provides biased output. One way of removing bias is described here https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-07 in Hashing to a Finite Field

**Review by shamatar — COMMENTED** [review]
I'd also want to see this https://medium.com/blockstream/insecure-shortcuts-in-musig-2ad0d38a97da implemented (so you commit to R points before the protocol, and that scheme in general follows this structure of rounds)

File: Cargo.toml
Comment (SanjoDeundiak) [reply Cargo.toml]: Done

File: src/musig.rs
Comment (SanjoDeundiak) [reply src/musig.rs]: Hash computation should be at least 100-1000x slower than dynamic dispatch call, so I decided not to use template parameters just to make code look cleaner (5 template parameters is harsh)

File: src/musig.rs
Comment (SanjoDeundiak) [reply src/musig.rs]: Done

File: README.md
Comment (SanjoDeundiak) [reply README.md]: It depends on Markdown engine. If we're talking about GitHub - you're right, originally I used Clion for creating this, which expected rs to be used
P.S. I also fixed js->javascript

File: src/musig.rs
Comment (SanjoDeundiak) [reply src/musig.rs]: It's not possible while not using templates for the class itself

File: src/musig.rs
Comment (SanjoDeundiak) [reply src/musig.rs]: This is rather controversial. What do you mean by "not used as a PrivateKey"? Any non-zero Fs element can be a PrivateKey, and there are a lot of possible usages for PrivateKey. Does the fact that we are multiplying it by generator means that we're extracting public key from private, or are we just performing multiplication on Fs element and curve point?
Should we call it PrivateKey? Fs? Maybe create separate struct? It's a matter of taste

File: src/musig.rs
Comment (SanjoDeundiak) [reply src/musig.rs]: Fixed

File: src/musig.rs
Comment (SanjoDeundiak) [reply src/musig.rs]: Fixed

File: src/musig_hash.rs
Comment (SanjoDeundiak) [reply src/musig_hash.rs]:
```
Is it possible that one will want to use AggregateHash but totally not interested in CommitmentHash in the context of this library?
```
It's not, but it doesn't mean we should combine all of the traits into one. Because:
1. It breaks SOLID principles
2. It would require us to create adapters for every combination of hashes we want to use (e.g. using musig with SHA512 + rescue and SHA512 + pedersen will require to create two structs on calling side with corresponding trait implementation)
3. This was a try to fix existing Franklin crypto problem with creating separate copy&paste functions for musig (e.g. musig_sha256_sign, musig_pedersen_sign, musig_rescue_sign) which would become even worse in case of using several hashes per algorithm like in this case.

File: tests/musig_tests.rs
Comment (SanjoDeundiak) [reply tests/musig_tests.rs]: Fixed

File: src/lib.rs
Comment (SanjoDeundiak) [reply src/lib.rs]: For some reason it doesn't work for me without this lines. I can investigate it, though I could be not the only one for whom this doesn't work, so I would leave it as is

File: tests/musig_tests.rs
Comment (SanjoDeundiak) [reply tests/musig_tests.rs]: This code is used to borrow one element of a slice as mutable, and others as immutable. Do you know how to do this better?

File: tests/musig_tests.rs
Comment (SanjoDeundiak) [reply tests/musig_tests.rs]: Fixed

File: tests/musig_tests.rs
Comment (SanjoDeundiak) [reply tests/musig_tests.rs]: Fixed

File: src/musig_wasm.rs
Comment (SanjoDeundiak) [reply src/musig_wasm.rs]: Fixed

File: src/musig_wasm.rs
Comment (SanjoDeundiak) [reply src/musig_wasm.rs]: Fixed

File: src/musig_hash.rs
Comment (SanjoDeundiak) [reply src/musig_hash.rs]: Fixed

File: src/musig_hash.rs
Comment (SanjoDeundiak) [reply src/musig_hash.rs]: Fixed

File: musig/src/hash.rs
Comment (popzxc) [inline musig/src/hash.rs:11]: > It's not, but it doesn't mean we should combine all of the traits into one. 

I don't quite agree. From the point of view of the main entity of this library -- multisig processor -- the suggested combinator trait serves exactly one abstraction: a complete set of hashing functions required for the multisig to operate.

This abstraction is still flexible enough on the part of the interface implementation. One can easily create a configurable implementor of this trait, just by using a template structure.

As a compromise, we can use both of your and my approaches: we will introduce one trait for `MusigSession`, e.g. `MusigSessionHasher`, and create a template-based implementor, like:

```rust
pub struct ConfigurableMusigHasher<AH, CH, SH, MH, E>
where
    A: AggregateHash<E>,
    C: CommitmentHash<E>,
    S: SignatureHash<E>,
    M: MsgHash,
    E: JubJubEngine,
 { .. }
```

With such a type, we will be able to type shortcuts, like:

```rust
pub type AltJubjubSha256MusigHasher = ConfigurableHasher<Sha256HStar, Sha256HStar, Sha256HStar, Sha256HStar, alt_babyjubjub::Bn256>;
```

Here is [a complete example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d1526212d861ed8c0d15624eb9160f55) of the similar snippet.

----

This may look a bit too verbose, but it achieves the multiple goals:

1. We don't lose any of the flexibility achieved by your approach.
2. We reduce amount of arguments required by the `MusigSession` constructor, reducing the visual complexity.
3. We will be able to use static dispatch instead of dynamic **while** not making the `MusigSession` awkward with too many template arguments.
4. By providing the type aliases for the most common hasher combinations, we won't burden the users with the template configuration of our `ConfigurableHasher`, unless they are not require this configuration. And even if they do, all they have to do is to create one similar type alias.

File: musig/src/musig.rs
Comment (popzxc) [inline musig/src/musig.rs]: >Hash computation should be at least 100-1000x slower than dynamic dispatch call, so I decided not to use template parameters just to make code look cleaner (5 template parameters is harsh)

With an approach I suggested above, two template parameters will suffice.

File: musig/src/musig.rs
Comment (popzxc) [inline musig/src/musig.rs:121]: > This is rather controversial. What do you mean by "not used as a PrivateKey"? Any non-zero Fs element can be a PrivateKey, and there are a lot of possible usages for PrivateKey. Does the fact that we are multiplying it by generator means that we're extracting public key from private, or are we just performing multiplication on Fs element and curve point?
> Should we call it PrivateKey? Fs? Maybe create separate struct? It's a matter of taste

Even though `PrivateKey` and `Fs` derive the same properties, they are extracted to the separate types for a reason. Let me explain by an analogy: `std::path::Path` and `std::str` derive the same properties, and any string can be a `Path`. But if you will convert a string to a path just to check if it doesn't contain a slash inside by using `is_absolute` method, it will be rather confusing.

This example is quite absurd, but I hope it shows my point: we must not use the type for anything but than its primary goal. `PrivateKey` objects must represent private keys, and otherwise it will lead to a confusion.

If there is a good reason to use `PrivateKey` type here, and it overweights the more clear (from the reader's perspective of view) solutions, I'd at least expect the good explanation of what's going on here and why it's done this way and not the other.

File: musig/tests/musig_tests.rs
Comment (popzxc) [inline musig/tests/musig_tests.rs]: >This code is used to borrow one element of a slice as mutable, and others as immutable. Do you know how to do this better?

I think it's an [XY problem](https://en.wikipedia.org/wiki/XY_problem). You don't really **have** to borrow mutable/immutable at the same time.

You can simply clone the data from the immutable element, and then iterate through other elements as mutable, skipping one of them, like:

```rust
let data = elemets[i].data().clone();

for (_, element) in elements.iter_mut().enumerate().filter(|(id, _el) id != i) {
    element.do_something(&data);
}
```

File: musig_wasm/src/lib.rs
Comment (popzxc) [inline musig_wasm/src/lib.rs]: >For some reason it doesn't work for me without this lines. I can investigate it, though I could be not the only one for whom this doesn't work, so I would leave it as is

In 2018 edition, you have to import macros just like any other items, e.g.:

[Pre-2018 approach](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=eb187f5c282ca48b81ff91874855bb3e) vs [2018 approach](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a340d1067edf6471b8522563514d4f87).

File: musig/src/musig.rs
Comment (popzxc) [inline musig/src/musig.rs:31]: Instead of having `performed_sign` as a flag, you can simply make `sign` method consuming the `self` by value, thus the whole object won't be usable once the signing operation is performed.

**Review by popzxc — COMMENTED** [review]
Thanks for your fixes and opinions on my comments! They are reasonable and motivated.

Since I (obviously) cannot mark my comments as resolved, I'd rather provide a brand new review, even though most of topics are continuation of the dialogue, just not to mix things up.

In general, currently PR looks much better. However, AFAICS not all the comments were addressed (e.g. I'd still love to see more tests and documentation in code). Do you have plans to address them later?

File: musig/tests/musig_tests.rs
Comment (SanjoDeundiak) [reply musig/tests/musig_tests.rs]: Yes, cloning elements was an obvious solution for the problem, but I wanted to avoid that. Borrowing some elements of a slice as mutable and others as immutable is totally legit and doesn't break any rust ideology. It's just the compiler that cannot figure out at compilation time that those two sets of elements doesn't intersect and we don't produce more than 1 reference for any of the slice elements. Solving this case was fun, but ok, I can just use clone.

File: musig_wasm/src/lib.rs
Comment (SanjoDeundiak) [reply musig_wasm/src/lib.rs]: Fixed

File: musig/src/musig.rs
Comment (SanjoDeundiak) [reply musig/src/musig.rs:31]: That will require to create separate struct for aggregating signature, since after creating a signature MusigSession aggregates signatures from other participants

File: musig/src/musig.rs
Comment (SanjoDeundiak) [reply musig/src/musig.rs:121]: Not all std::str can be a valid Path, so Path also provides some guarantees about what data it holds. When we're talking about Fs, PrivateKey and Seed they all hold exactly same data and the only difference is semantics and additional functions each of them has.
The author of original paper struggles himself to describe what this variable is, and he calls it just r and R. I just decided to use PrivateKey because:
1. It looks to me like ephemeral private key
2. We need R = r*G which also looks like ephemeral key pair and we have convenient function for that
3. It's unclear to me how changing type to Seed or Fs will help reader understand what is going on in this code

File: musig/tests/musig_tests.rs
Comment (popzxc) [reply musig/tests/musig_tests.rs]: Using `unsafe` in Rust is a sensitive topic. In general, there are two main reason (unrelated to environment, e.g. embedded) to use `unsafe`:

1. FFI.
2. Huge performance boost unachievable with safe code only.

Otherwise, it is highly recommended to avoid `unsafe`. Yeah, `unsafe` allows us to write code based on the data unavailable to compiler, but at the same time the **greatest** thing in Rust **is** the compiler, which makes you (not personally you, of course) worry about invariants of your code less. Each `unsafe` block in your code is a burden, which you have to carry. You'll have to maintain the invariants by your own, since you've told the compiler that you got it.

Thus, a small optimization here in my opinion does not overweight the introduced inconsistency.

If you are interested in topic, you can read about the usage of `unsafe` in `actix-web`: it was a big discussion (even close to holy war / flame) about `actix-web` library using `unsafe` without a strong motivation. While in most cases the crate maintainer definitely knew what he's doing, some of the code turned out to be unsound. At the same time, it has lead to a reputational damage, just because it's not conventional in Rust community to discard the guarantees (and indeed, without these guarantees, Rust is just *yet another language*).

File: musig/src/musig.rs
Comment (popzxc) [reply musig/src/musig.rs:31]: Ah, I see, sorry. My bad.

File: musig/tests/musig_tests.rs
Comment (SanjoDeundiak) [reply musig/tests/musig_tests.rs]: Everything related to iterations in std::slice is implemented using unsafe, so I don't see any problem in writing one more split function. What are other options? Copy data whenever you don't find corresponding standard function? No matter how big data is? Or create pull request to std lib and wait for approval?

File: src/musig_hash.rs
Comment (SanjoDeundiak) [reply src/musig_hash.rs]: Since different hashes provide output of different size how would you do that? More templates?

File: musig/tests/musig_tests.rs
Comment (popzxc) [reply musig/tests/musig_tests.rs]: Code in `std` is allowed to be `unsafe` simply because standard library is under the constant supervision of the many qualified programmers, which work hard to ensure that every use of `unsafe` is correct and unavoidable.

I'm not saying that `unsafe` is an evil creature which is forbidden to be used. It **has** its usages and the performance is one of them (and this is stated in my comment).

However, performance of the test code is not the key priority, and if it's about choosing either of the unsafe code or data copying, I'd choose the latter.

P.S. Regarding

>  Copy data whenever you don't find corresponding standard function

I'm by no means an expert Rust programmer which has seen every possible case, but I've spent the last few years writing 95% of my code in Rust, both for the production and personal projects. However, cases where I clearly had to use `unsafe` are not numerous, and the majority of them were caused by FFI. So I don't really agree that it's always about either of `unsafe` or data copying when you can't find corresponding standard function.

File: musig/src/musig.rs
Comment (popzxc) [reply musig/src/musig.rs:121]: Okay, I see your point. As a compromise I can suggest a solution which probably will help the reader more than the type name:

Let's add a comment explaining what's going on in this code. A couple of words about how `PrivateKey` is used to calculate `R` and what this `R` has to do with the signing process will reduce any kind of possible misunderstanding.

Is this OK for you?

Comment (SanjoDeundiak) [reply musig/src/musig.rs:121]: I think restoring algorithm from source code is not the best idea considering you have a paper that describes it. And if one read the paper, there's no need to describe what r is. From the other hand developers who doesn't want to dive into how this works can just follow readme and not think what r, t and other variables mean.

File: musig/tests/musig_tests.rs
Comment (SanjoDeundiak) [reply musig/tests/musig_tests.rs]: Fixed

File: musig/src/musig.rs
Comment (SanjoDeundiak) [reply musig/src/musig.rs]: Fixed

File: musig/src/hash.rs
Comment (SanjoDeundiak) [reply musig/src/hash.rs:11]: Fixed

File: musig/src/musig.rs
Comment (popzxc) [reply musig/src/musig.rs:121]: Sure, then let it be as is.

File: musig/src/hash.rs
Comment (SanjoDeundiak) [reply musig/src/hash.rs:11]: BTW because of changing this to what you suggested it's now impossible to change hash type in runtime from wasm. It can now use only one hash and it should be predefined

Comment (popzxc) [reply musig/src/hash.rs:11]: If one will want to provide multiple hashers for the WASM, they'll be able to create one more crate with required predefined hashers. And since the hash set is a somewhat stable topic, I think it's OK.

File: src/musig_hash.rs
Comment (SanjoDeundiak) [reply src/musig_hash.rs]: Unfortunately Franklin crypto types don't implement Debug, so I only derived Debug for types that don't aggregate Franklin crypto types (there were only few)

File: src/musig_hash.rs
Comment (SanjoDeundiak) [reply src/musig_hash.rs]: That's exactly what TODO: optimise in place of using meant.
Fixed.

File: src/musig_wasm.rs
Comment (SanjoDeundiak) [reply src/musig_wasm.rs]: Fixed

File: tests/musig_wasm_tests.rs
Comment (SanjoDeundiak) [reply tests/musig_wasm_tests.rs]: This may be overkill, but since this code is not autogenerated there could be mistakes and I would prefer to have functional test for it. Also, I'll add more tests. I even have JS test which is gitignored, though I should probably push it

File: .gitignore
Comment (SanjoDeundiak) [reply .gitignore]: Fixed

File: src/musig_hash.rs
Comment (SanjoDeundiak) [reply src/musig_hash.rs]: Fixed

File: src/musig_hash.rs
Comment (popzxc) [reply src/musig_hash.rs]: Nah, it was more of a question, it's OK as is.

File: src/musig.rs
Comment (SanjoDeundiak) [reply src/musig.rs]: Fixed

File: tests/musig_tests.rs
Comment (SanjoDeundiak) [reply tests/musig_tests.rs]: Fixed

**Review by popzxc — APPROVED** [review]
Looks good, thanks!

**Pull request closed without merging**
